tfx_addons/sampler/executor.py

```diff
@@ -0,0 +1,259 @@
+                side=0):
+  """Function called in a Beam pipeline that performs sampling using Python's
+  random module on an input key:value pair, where the key is the class label
+  and the values are the data points to sample. Note that the key is discarded."""
+
+  if sampling_strategy == spec.SamplingStrategy.UNDERSAMPLE:
+    random_sample_data = random.sample(val, side)
+  elif sampling_strategy == spec.SamplingStrategy.OVERSAMPLE:
+    random_sample_data = random.choices(val, k=side)
+
+  for item in random_sample_data:
+    yield item
+
+
+def filter_null(item, keep_null=False, null_vals=None):
+  """Function that filters all of the null labels (and any optional labels
+  from the dataset at large. Returns either the null-labeled values or the
+  non-null-labeled values, depending on the value of keep_null."""
+
+  if item[0] == 0:
+    keep = True
+  else:
+    keep = bool(item[0])
+
+  if null_vals and str(item[0]) in null_vals and keep:
+    keep = False
+  keep ^= keep_null
+
+  if keep:
+    return item
+  else:
+    return None
+
+
+def sample(p, uri, label, shards, null_classes, output_dir, sampling_strategy):
+  """Function that actually samples the given split.
+
+  Args:
+    uri: The input uri for the specific split of the input example artifact.
+    label: The name of the column containing class names to sample by.
+    shards: The number of files that each sampled split should
+    contain. Default 0 is Beam's tfrecordio function's default.
+    null_classes: A list determining which classes that we should
+    not sample. Defaults to None.
+    output_dir: The output directory for the split of the output artifact.
+  Returns:
+    None
+  """
+
+  data = read_tfexamples(p, uri, label)
+  merged = sample_examples(data, null_classes, sampling_strategy)
+  write_tfexamples(merged, shards, output_dir)
+
+
+def read_tfexamples(p, uri, label):
+  """Function that reads tf.Examples from tfRecord files and converts them
+  to a K-V PCollection usable by Beam."""
+
+  dataset = tf.data.TFRecordDataset(tf.data.Dataset.list_files(f'{uri}/*'),
+                                    compression_type="GZIP")
+
+  # Take the input TFRecordDataset and extract the class label that we want.
+  # Output format is a K-V PCollection: {class_label: TFRecord in string format}
+  data = (p
+          | "DatasetToPCollection" >> beam.Create(dataset)
+          | "MapToLabel" >> beam.Map(generate_elements, label))
+  return data
+
+
+def sample_examples(data, null_classes, sampling_strategy):
+  """Function that performs the sampling given a label-mapped dataset."""
+
+  # Finds the minimum frequency of all classes in the input label.
+  # Output is a singleton PCollection with the minimum # of examples.
+
+  def find_minimum(elements):
+    return min(elements or [0])
+
+  def find_maximum(elements):
+    return max(elements or [0])
+
+  if sampling_strategy == spec.SamplingStrategy.UNDERSAMPLE:
+    sample_fn = find_minimum
+  else:
```

Comment:
We implicitly assume `SamplingStrategy` has only 2 enum values here (and that the other is definitely oversample). I would make that assumption explicit: `elif... OVERSAMPLE` and then `else:... raise ValueError`